### ISSUE TITLE
Rule proposal: prefer `location.assign()` over setting `location.href`

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#2944]
### Description

At first I thought I was against this, but I think it makes sense:

- `location.assign()` matches `location.replace()` in style, so it more consistently changes the URL
- setting `location.href` is actually "not great" if you think that it has a side effect, a pretty big one at that. 

### Examples

```js
// ❌
location.href = '/'

// ✅
location.assign('/')
```

### Proposed rule name

prefer-location-assign

### Additional Info

The drawback is that people are less aware of what the method does, but that's the case for any better yet lesser known APIs, like Temporal over Date, toLocaleString over toString, etc

**Comment by silverwind** [comment]
Useful. Also make it work on `window.location.href` and `globalThis.location.href`.

Maybe even generalize into a `no-legacy-setter-apis`, browser have a few legacy APIs like that which should not be used if they have better alternatives available.

**Comment by fregante** [comment]
I forgot to mention https://github.com/sindresorhus/eslint-plugin-unicorn/issues/402, which was closed due to lack of suggestions. We need a list of bad APIs before making this a generic rule

**Comment by silverwind** [comment]
I'll try to provide a comprehensive list later.

[`process.title`](https://nodejs.org/api/process.html#processtitle) in node is another api that could eventually be covered by a general rule once https://github.com/nodejs/node/issues/62797 is fixed, but it might be too early to include because it will likely only be in Node.js v27+.

**Comment by silverwind** [comment]
Here is a list of web APIs that could be covered in a `no-legacy-setter-apis` (name to be decided), all have strictly equivalent replacements. In Node, I didn't find any API's besides `process.title` that matches the criteria.

- [`Location.href`](https://developer.mozilla.org/en-US/docs/Web/API/Location/href) → [`Location.assign()`](https://developer.mozilla.org/en-US/docs/Web/API/Location/assign)
- [`Element.scrollTop`](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollTop) → [`Element.scrollTo()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollTo)
- [`Element.scrollLeft`](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollLeft) → [`Element.scrollTo()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollTo)
- [`Element.className`](https://developer.mozilla.org/en-US/docs/Web/API/Element/className) → [`Element.setAttribute()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute) (debatable)
- [`CSSStyleDeclaration.cssText`](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/cssText) → [`Element.setAttribute()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute) (debatable)

**Comment by fregante** [comment]
I don’t think they're legacy, also unfortunately they're not equivalent. In order to use scrollTo() to set the scrollTop (the most common case) you must provide scrollLeft as well. 

setAttribute('class', n) would  make sense but it's longer so it might receive more pushback. It also goes the opposite way of the `prefer-dom-node-dataset` rule

**Comment by silverwind** [comment]
Yeah legacy is not a good name, they are not deprecated, they just have more ergonomic alternatives.

So with the scroll not being a simple replacement (maybe single direction works with object form), the only real candidates that are left are `location.href` and `process.title` (in the far future).

**Comment by silverwind** [comment]
Can we agree on `no-setter-apis` name? Willing to implement this with just `location.href` for now, to be extended later.

**Comment by fregante** [comment]
I'll let the maintainers decide, but to me that sounds too generic. Without other examples it's not clear that the rule will be limited to setters either. 

I'd start with the name I originally suggested and then *if* more are needed the rule can be renamed. In reality this is a super simple rule so if more are needed they can be separate rules. *If* process.setTitle is ever added, Unicorn can't support it for a couple of years anyway.

**Comment by silverwind** [comment]
I still prefer forward-thinking names. Better have a good name now then more work later. A rule rename is always disruptive to users.

**Comment by silverwind** [comment]
I guess the benefit of a targeted rules is to enable user more fine-grained control over the affected APIs.

BTW this rule could almost be implemented in https://eslint.org/docs/latest/rules/no-restricted-properties, but unfortunately, that rule has no way of flagging only write access.
